### PR TITLE
Track which fibers scheduled the current render work

### PR DESCRIPTION
**Pull request by bvaughn**
This PR introduces the concept of tracking which Fiber(s) schedule work with React.

Tracked Fibers are called "updaters" and are exposed to DevTools via a `memoizedUpdaters` property on the `ReactFiberRoot`. The implementation of this feature follows a vaguely similar approach as interaction tracing, but does not require reference counting since there is no subscriptions API.

This change is in support of a new DevTools Profiler feature that shows which Fiber(s) scheduled the selected commit in the Profiler.

All changes have been gated behind a new feature flag, `enableUpdaterTracking`, which is enabled for Profiling builds by default. We also only track updaters when DevTools has been detected, to avoid doing unnecessary work.

### CODE REVIEW

**Comment by bvaughn** [comment]
Hm. I think there's a gap here with flushing pending updates that can cause the updater to get lost. Haven't been able to catch it in a test yet.

**Comment by bvaughn** [comment]
Okay. I think I captured a variant of the bug in a new test (failing before) and fixed.

**Comment by bvaughn** [comment]
Rebased after #15664 landed.

**Comment by acdlite** [comment]
I'm going to defer to @sebmarkbage to review this since I think you discussed the overall approach with him already

**Comment by bvaughn** [comment]
Lucky @sebmarkbage 🎉

**Comment by bvaughn** [comment]
Ping @sebmarkbage 😄

**Comment by bvaughn** [comment]
Ping @sebmarkbage again 👋

**Comment by bvaughn** [comment]
I think this PR is still worth at least considering. Leaving open for now until @sebmarkbage indicates level of interest. (If interested, I'll rebase.)

**Comment by bvaughn** [comment]
This would have been a useful feature 😄  It's unfortunate that it never got reviewed.

**Comment by gaearon** [comment]
I can review it! I don't think I've seen it before. I doubt Seb sees GH mentions. :)

**Comment by bvaughn** [comment]
> I don't think I've seen it before.

Screen Shot 2021-03-29 at 10 24 26 PM

🤣  At least I'm not the only one who does this!

Thanks for the offer Dan! I'll clean it up sometime this week and ping you.

**Comment by gaearon** [comment]
It’s been... a while.

**Comment by bvaughn** [comment]
Well fortunately, only....everything...has changed since I wrote this. Should not be a problem.

**Comment by bvaughn** [comment]
I'm returning this to draft status while I work on rebasing and updating the expiration times logic to lanes logic.

File: packages/react-reconciler/src/ReactFiberLane.new.js
Comment (bvaughn) [inline packages/react-reconciler/src/ReactFiberLane.new.js:796]: It seemed best to put these functions in `ReactFiberLane` so they had access to `laneToIndex` to make lanes->indices mapping easier.

**Comment by bvaughn** [comment]
Okay I think it's ready for a look 😄 

(I assume I'l have to add some gating to my test once CI shows me which builds fail but otherwise...)

**Comment by sebmarkbage** [comment]
Didn't see pings on GH.

I think I must have forgotten a lot of the context here.

But as I read it, this is really mostly because we want to track it through retries of Suspense and Offscreen. Is that right?

Because otherwise we can infer this info in the devtools when we loop over the tree after a commit. If it's just for that commit.

This is similar to how the Cache has to attach itself to a Suspense boundary.

I wonder if this would be more scoped if we did it that way. We could store the parents that participated in an update on an internal context and then snapshot that on the Suspense/Offscreen/Hydrating boundary.

Then devtools would pick up this info in the traversal of the commit either from the updated hook or the suspense boundary that was retries.

There's a subtle difference also if you create this list when something is called with setState on it vs if it participated in the render. Because things that called setState but that thing was deleted or in a hidden subtree didn't contribute to the rendering (just the cloning) of that commit. Not sure which one is better.

File: packages/react-reconciler/src/ReactFiberCommitWork.new.js
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberCommitWork.new.js:2100]: It would be better to store this on a module variable so we don't have to pay the cost of the extra writing to the stack in prod.

**Comment by bvaughn** [comment]
> But as I read it, this is really mostly because we want to track it through retries of Suspense and Offscreen. Is that right?

Yes.

> I wonder if this would be more scoped if we did it that way. We could store the parents that participated in an update on an internal context and then snapshot that on the Suspense/Offscreen/Hydrating boundary.
>
> Then devtools would pick up this info in the traversal of the commit either from the updated hook or the suspense boundary that was retries.

Some initial concern at the idea of maintaining that over time, across React versions. Seems like the internal logic of Suspense (and Offscreen) has already changed a lot and will likely be changing more (e.g. with effects semantics) – something explicit feels more future proof.

File: packages/react-reconciler/src/ReactFiberCommitWork.new.js
Comment (bvaughn) [reply packages/react-reconciler/src/ReactFiberCommitWork.new.js:2100]: Sure.

Seems a little arbitrary, since we're already passing around at least the root to most of these functions, but I'll make the minimal change for now and use module scope for the root and lanes.

Comment (bvaughn) [reply packages/react-reconciler/src/ReactFiberCommitWork.new.js]: Done.

**Comment by bvaughn** [comment]
CI failure is unrelated to code changes:
> Error response from daemon: Get https://registry-1.docker.io/v2/circleci/openjdk/manifests/8-jdk-node-browsers: Get https://auth.docker.io/token?account=circlecidockerdefaultreadonly&scope=repository%3Acircleci%2Fopenjdk%3Apull&service=registry.docker.io: EOF

**Comment by bvaughn** [comment]
Can I get a review of this please? :)

**Review by sebmarkbage — APPROVED** [review]
We can discuss alternative approaches for a longer term solution but don't want to block the UI features.

**Comment by bvaughn** [comment]
Much appreciated!

**Pull request merged**
